README.md
RAUC hawkBit Updater
====================

[![Build Status](https://github.com/rauc/rauc-hawkbit-updater/workflows/tests/badge.svg)](https://github.com/rauc/rauc-hawkbit-updater/actions)
[![License](https://img.shields.io/badge/license-LGPLv2.1-blue.svg)](https://raw.githubusercontent.com/rauc/rauc-hawkbit-updater/master/LICENSE)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/rauc/rauc-hawkbit-updater.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/rauc/rauc-hawkbit-updater/alerts/)
[![Language grade: C/C++](https://img.shields.io/lgtm/grade/cpp/g/rauc/rauc-hawkbit-updater.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/rauc/rauc-hawkbit-updater/context:cpp)

The RAUC hawkBit updater is a simple commandline tool / daemon written in C (glib).
It is a port of the RAUC hawkBit Client written in Python.
The daemon runs on your target and operates as an interface between the
[RAUC D-Bus API](https://github.com/rauc/rauc)
and the [hawkBit DDI API](https://github.com/eclipse/hawkbit).

Quickstart
----------

The RAUC hawkBit updater is primarily meant to be used as a daemon,
but it also allows you to do a one shot instantly checking and install
new software.

To quickly getting started with hawkBit server, follow
[this](https://github.com/eclipse/hawkbit#getting-started)
instruction.

Setup target (device) configuration file:

```ini
  [client]
  hawkbit_server            = 127.0.0.1:8080
  ssl                       = false
  ssl_verify                = false
  tenant_id                 = DEFAULT
  target_name               = test-target
  auth_token                = bhVahL1Il1shie2aj2poojeChee6ahShu
  #gateway_token            = bhVahL1Il1shie2aj2poojeChee6ahShu
  bundle_download_location  = /tmp/bundle.raucb
  retry_wait                = 60
  connect_timeout           = 20
  timeout                   = 60
  log_level                 = debug

  [device]
  product                   = Terminator
  model                     = T-1000
  serialnumber              = 8922673153
  hw_revision               = 2
  key1                      = value
  key2                      = value
```

All key/values under [device] group are sent to hawkBit as data (attributes).
The attributes in hawkBit can be used in target filters.

Finally start the updater as daemon:

```shell
  ./rauc-hawkbit-updater -c config.conf
```


Debugging
---------

When setting the log level to 'debug' the RAUC hawkBit client will print
JSON payload sent and received. This can be done by using option -d.

```shell
  ./rauc-hawkbit-updater -d -c config.conf
```


Compile
-------

```shell
  mkdir build
  cd build
  cmake ..
  make
```


Usage / options
---------------

```shell
/usr/bin/rauc-hawkbit-updater --help
Usage:
  rauc-hawkbit-updater [OPTION?]

Help Options:
  -h, --help               Show help options

Application Options:
  -c, --config-file        Configuration file
  -v, --version            Version information
  -d, --debug              Enable debug output
  -r, --run-once           Check and install new software and exit
  -s, --output-systemd     Enable output to systemd
```


CHANGES
Release 0.1 (in development)
----------------------------

This is the initial release of RAUC hakBit Updater.


test/test.py
import pytest
import sys, os, subprocess
import re
import logging

import hawkbit
from hawkbit import HawkbitTestClient
from hawkbit import HawkbitError

from tempfile import NamedTemporaryFile

here = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) + "/build"
os.environ["PATH"] = here + ":" + os.environ["PATH"]

@pytest.fixture(scope="session")
def hawkbit_configured():
    hawkbit = HawkbitTestClient("localhost", "8080", username="admin", password="admin")

    filename_a = os.path.dirname(os.path.abspath(__file__)) + "/dummy-bundle.raucb"

    # Set config parameters
    hawkbit.set_config("pollingTime", "00:00:30")
    hawkbit.set_config("pollingOverdueTime", "00:03:00")
    
    # Add expected target
    hawkbit.add_target("test-controller", "bhVahL1Il1shie2aj2poojeChee6ahShu")

    # Create modules with artifacts
    module_id_a = hawkbit.add_swmodule('Test module')
    hawkbit.add_artifact(module_id_a, filename_a)

    # Create distributions of it
    dist_id = hawkbit.add_distributionset(module_id_a, name="Poky-Test")
    logging.info("Added demo distributions")

    hawkbit.assign_target(dist_id, "test-controller")

def run(command):
    proc = subprocess.Popen(command,
            stdout = subprocess.PIPE,
            stderr = subprocess.PIPE,
            )
    out,err = proc.communicate()
    return out, err, proc.returncode

def test_rauc_hawkbit_version():
    command = ["rauc-hawkbit-updater", "-v"]
    out, err, exitcode = run(command)
    assert exitcode == 0
    assert re.match(b'Version ?.?', out)
    assert err == b''

def test_rauc_hawkbit_invalid_arg():
    command = ["rauc-hawkbit-updater", "--invalidarg"]
    out, err, exitcode = run(command)
    assert exitcode == 1
    assert out == b''
    assert err == b'option parsing failed: Unknown option --invalidarg\n'

def test_rauc_hawkbit_no_config_file():
    command = ["rauc-hawkbit-updater"]
    out, err, exitcode = run(command)
    assert exitcode == 2
    assert out == b''
    assert err == b'No configuration file given\n'

def test_rauc_hawkbit_non_existing_config_file():
    command = ["rauc-hawkbit-updater", "-c", "does-not-exist.conf"]
    out, err, exitcode = run(command)
    assert exitcode == 3
    assert out == b''
    assert err == b'No such configuration file: does-not-exist.conf\n'

def test_rauc_hawkbit_register_and_check_valid_auth(hawkbit_configured):
    with NamedTemporaryFile(mode="w", delete=False) as tmp_source:
        with open("test-config.conf") as source:
            for line in source.readlines():
                tmp_source.write(re.sub(r'TEST_TOKEN', r'bhVahL1Il1shie2aj2poojeChee6ahShu', line))
    command = ["rauc-hawkbit-updater", "-c", tmp_source.name, "-r"]
    print(command)
    out, err, exitcode = run(command)
    assert exitcode == 0
    assert re.match(b'MESSAGE: Checking for new software....*', out)
    assert err == b''

def test_rauc_hawkbit_register_and_check_invalid_auth(hawkbit_configured):
    with NamedTemporaryFile(mode="w", delete=False) as tmp_source:
        with open("test-config.conf") as source:
            for line in source.readlines():
                tmp_source.write(re.sub(r'TEST_TOKEN', r'ahVahL1Il1shie2aj2poojeChee6ahShu', line))
    command = ["rauc-hawkbit-updater", "-c", tmp_source.name, "-r"]
    print(command)
    out, err, exitcode = run(command)
    assert exitcode == 1
    assert re.match(b'MESSAGE: Checking for new software....*', out)
    assert err == b'CRITICAL: Failed to authenticate. Check if auth_token is correct?\n'


test/hawkbit.py
import attr
import requests as r

from json.decoder import JSONDecodeError

@attr.s(eq=False)
class HawkbitTestClient:
    host = attr.ib(validator=attr.validators.instance_of(str))
    port = attr.ib(validator=attr.validators.instance_of(str))
    username = attr.ib(validator=attr.validators.instance_of(str))
    password = attr.ib(validator=attr.validators.instance_of(str))
    version = attr.ib(
        default=1.0, validator=attr.validators.instance_of(float)
    )

    def __attrs_post_init__(self):
        self.url = 'http://{host}:{port}/rest/v1/{endpoint}'

    def set_config(self, key, value):
        self.put('system/configs/{}'.format(key), {"value" : value})

    def add_target(self, target_id: str, token: str):
        """Add a target to the HawkBit Installation
        Arguments:
          - target_id(str): the (unique) device name of the target to add
          - token(str): pre-shared key to authenticate the target
        """
        testdata = [{
            'controllerId': target_id,
            'name': target_id,
            'description': 'test',
            'securityToken': token
        }]

        self.post_json('targets', testdata)

    def add_target_filter(self, name: str, query: str):
        testdata = {
            'name': name,
            'query' : query
        }

        response = self.post_json('targetfilters', testdata)
        return respone['id']

    def delete_target(self, target_id: str):
        """Delete a target from the HawkBit Installation
        Arguments:
          - target_id(str): the (unique) device name of the target to delete
        """

        self.delete('targets/{}'.format(target_id))

    def add_swmodule(self, modulename: str):
        testdata = [{
            'name': modulename,
            'version': str(self.version),
            'type': 'os'
        }]

        self.module_id = self.post_json('softwaremodules', testdata)[0]['id']
        return self.module_id

    def delete_swmodule(self, module_id: str):
        """Delete a softwaremodule from the HawkBit Installation
        Arguments:
          - module_id(str): the ID given by hawkBit for the module
        """

        self.delete('softwaremodules/{}'.format(module_id))

    def add_distributionset(self, module_id, name=None):
        testdata = [{
            'name': name if name else 'Test_distribution {}'.format(self.version),
            'description': 'Test distribution',
            'version': str(self.version),
            'modules': [{
                'id': module_id
            }],
            'type': 'os'
        }]

        self.distribution_id = self.post_json('distributionsets',
                                              testdata)[0]['id']
        return self.distribution_id

    def get_distribution_id(self, distname):
        distributionsets = self.get_endpoint('distributionsets')
        for dist in distributionsets['content']:
            if dist['name'] == distname:
                return dist['id']
        return None

    def delete_distributionset(self, distset_id: str):
        """Delete a distrubitionset from the HawkBit Installation
        Arguments:
          - distset_id(str): the ID of the distribution set to delete
        """

        self.delete('distributionsets/{}'.format(distset_id))

    def add_artifact(self, module_id: str, filename: str):
        endpoint = 'softwaremodules/{}/artifacts'.format(module_id)

        return self.post_binary(endpoint, filename)['id']

    def delete_artifact(self, module_id: str, artifact_id: str):
        """Delete an artifact from the HawkBit Installation
        Arguments:
          - artifact_id(str): the ID of the artifact to delete
        """

        self.delete('softwaremodules/{}/artifacts/{}'.format(module_id, artifact_id))

    def assign_target(self, distribution_id, target_id):
        endpoint = 'distributionsets/{}/assignedTargets'.format(
            distribution_id
        )
        testdata = [{'id': target_id}]

        self.post_json(endpoint, testdata)
        # Increment version to be able to flash over an already deployed distribution
        self.version = self.version + 0.1

    def add_rollout(self, name, distribution_id, groups, target_filter="id==*"):
        # TODO: filter by overdue, filter by other
        testdata = {
            'name': name,
            'distributionSetId': distribution_id,
            'targetFilterQuery': target_filter,
            'amountGroups': groups
        }

        self.rollout_id = self.post_json('rollouts', testdata)['id']
        return self.rollout_id

    def start_rollout(self, rollout_id):
        endpoint = 'rollouts/{}/start'.format(
            rollout_id
        )

        self.post(endpoint)

    def post(self, endpoint: str):
        req = r.post(
            self.url.format(
                endpoint=endpoint, host=self.host, port=self.port
            ),
            auth=(self.username, self.password),
        )
        if req.status_code != 200 and req.status_code != 201:
            raise HawkbitError(
                'Wrong statuscode, got {} instead of 200/201'.
                format(req.status_code),
                req.status_code,
                req.json(),
            )

    def post_json(self, endpoint: str, data: dict):
        headers = {'Content-Type': 'application/json;charset=UTF-8'}
        req = r.post(
            self.url.format(
                endpoint=endpoint, host=self.host, port=self.port
            ),
            headers=headers,
            auth=(self.username, self.password),
            json=data
        )
        if req.status_code != 200 and req.status_code != 201:
            raise HawkbitError(
                'Wrong statuscode, got {} instead of 200/201, with error {}'.
                format(req.status_code, req.json()),
                req.status_code,
                req.json(),
            )
        return req.json()

    def post_binary(self, endpoint: str, filename: str):
        files = {'file': open(filename, 'rb')}
        req = r.post(
            self.url.format(
                endpoint=endpoint, host=self.host, port=self.port
            ),
            auth=(self.username, self.password),
            files=files
        )
        if req.status_code != 201:
            try:
                json_response = req.json()
            except JSONDecodeError:
                json_response = None
            raise HawkbitError(
                'Wrong statuscode, got {} instead of 201, with error {}'.
                format(req.status_code, json_response),
                req.status_code,
                json_response,
            )
        return req.json()

    def put(self, endpoint: str, data: dict):
        req = r.put(
            self.url.format(
                endpoint=endpoint, host=self.host, port=self.port
            ),
            auth=(self.username, self.password),
            json=data
        )
        if req.status_code != 200 and req.status_code != 201:
            raise HawkbitError(
                'Wrong statuscode, got {} instead of 200/201'.
                format(req.status_code),
                req.status_code,
                req.json(),
            )

    def delete(self, endpoint: str):
        req = r.delete(
            self.url.format(
                endpoint=endpoint, host=self.host, port=self.port
            ),
            auth=(self.username, self.password),
        )
        if req.status_code != 200:
            raise HawkbitError(
                'Wrong statuscode, got {} instead of 200, with error {}'.
                format(req.status_code, req.json()),
                req.status_code,
                req.json(),
            )

    def get_endpoint(self, endpoint: str):
        headers = {'Content-Type': 'application/json;charset=UTF-8'}
        req = r.get(
            self.url.format(
                endpoint=endpoint, host=self.host, port=self.port
            ),
            headers=headers,
            auth=(self.username, self.password)
        )
        if req.status_code != 200:
            raise HawkbitError(
                'Wrong statuscode, got {} instead of 200, with error {}'.
                format(req.status_code, req.json()),
                req.status_code,
                req.json(),
            )
        return req.json()


@attr.s(eq=False)
class HawkbitError(Exception):
    msg = attr.ib()
    status_code = attr.ib()
    json = attr.ib()


src/log.c
/**
 * SPDX-License-Identifier: LGPL-2.1-only
 *
 * Copyright (C) 2018-2020 Prevas A/S (www.prevas.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @file   log.c
 * @author Lasse Mikkelsen <lkmi@prevas.dk>
 * @date   19 Sep 2018
 * @brief  Log handling
 */

#include "log.h"

static gboolean output_to_systemd = FALSE;

/**
 * @brief convert GLogLevelFlags to string
 *
 * @param[in] level Log level that should be returned as string.
 */
static const gchar *log_level_to_string(GLogLevelFlags level)
{
        switch (level) {
        case G_LOG_LEVEL_ERROR:
                return "ERROR";
        case G_LOG_LEVEL_CRITICAL:
                return "CRITICAL";
        case G_LOG_LEVEL_WARNING:
                return "WARNING";
        case G_LOG_LEVEL_MESSAGE:
                return "MESSAGE";
        case G_LOG_LEVEL_INFO:
                return "INFO";
        case G_LOG_LEVEL_DEBUG:
                return "DEBUG";
        case G_LOG_FLAG_FATAL:
                return "FATAL";
        case G_LOG_FLAG_RECURSION:
                return "RECURSION";
        case G_LOG_LEVEL_MASK:
                return "MASK";
        default:
                return "UNKNOWN";
        }
}

/**
 * @brief     map glib log level to syslog
 *
 * @param[in] level Log level that should be returned as string.
 * @return    syslog level
 */
#ifdef WITH_SYSTEMD
static int log_level_to_int(GLogLevelFlags level)
{
        switch (level) {
        case G_LOG_LEVEL_ERROR:
                return LOG_ERR;
        case G_LOG_LEVEL_CRITICAL:
                return LOG_CRIT;
        case G_LOG_LEVEL_WARNING:
                return LOG_WARNING;
        case G_LOG_LEVEL_MESSAGE:
                return LOG_NOTICE;
        case G_LOG_LEVEL_INFO:
                return LOG_INFO;
        case G_LOG_LEVEL_DEBUG:
                return LOG_DEBUG;
        case G_LOG_FLAG_FATAL:
                return LOG_FATAL;
        case G_LOG_FLAG_RECURSION:
                return LOG_RECURSION;
        case G_LOG_LEVEL_MASK:
                return LOG_MASK;

        default:
                return LOG_INFO;
        }
}
#endif

/**
 * @brief     Glib log handler callback
 *
 * @param[in] log_domain Log domain
 * @param[in] log_level  Log level
 * @param[in] message    Log message
 * @param[in] user_data  Not used
 */
static void log_handler_cb(const gchar    *log_domain,
                           GLogLevelFlags log_level,
                           const gchar    *message,
                           gpointer user_data)
{
#ifdef WITH_SYSTEMD
        if (output_to_systemd) {
                int log_level_int = log_level_to_int(log_level & G_LOG_LEVEL_MASK);
                sd_journal_print(log_level_int, "%s", message);
        } else {
#endif
        const gchar *log_level_str;
        log_level_str = log_level_to_string(log_level & G_LOG_LEVEL_MASK);
        if (log_level <= G_LOG_LEVEL_WARNING) {
                g_printerr("%s: %s\n", log_level_str, message);
        } else {
                g_print("%s: %s\n", log_level_str, message);
        }
#ifdef WITH_SYSTEMD
}
#endif
}

/**
 * @brief     Setup Glib log handler
 *
 * @param[in] domain Log domain
 * @param[in] level  Log level
 * @param[in] p_output_to_systemd output to systemd journal
 */
void setup_logging(const gchar *domain, GLogLevelFlags level, gboolean p_output_to_systemd)
{
        output_to_systemd = p_output_to_systemd;
        g_log_set_handler(NULL,
                          level | G_LOG_FLAG_FATAL | G_LOG_FLAG_RECURSION,
                          log_handler_cb, NULL);
}


src/json-helper.c
/**
 * SPDX-License-Identifier: LGPL-2.1-only
 *
 * Copyright (C) 2018-2020 Prevas A/S (www.prevas.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @file json-helper.c
 * @author Lasse Mikkelsen <lkmi@prevas.dk>
 * @date 19 Sep 2018
 * @brief JSON helper functions
 *
 */

#include "json-helper.h"

gchar* json_get_string(JsonNode *json_node, const gchar *path)
{
        char *res_str = NULL;
        JsonNode *result = json_path_query(path, json_node, NULL);
        if (result) {
                JsonArray *results = json_node_get_array(result);
                if (json_array_get_length( results ) > 0) {
                        res_str = g_strdup(json_array_get_string_element(results, 0));
                }
        }
        json_node_unref(result);
        return res_str;
}

gint64 json_get_int(JsonNode *json_node, const gchar *path)
{
        gint64 res_int = -1;
        JsonNode *result = json_path_query(path, json_node, NULL);
        if (result) {
                JsonArray *results = json_node_get_array(result);
                if (json_array_get_length( results ) > 0) {
                        res_int = json_array_get_int_element(results, 0);
                }
        }
        json_node_unref(result);
        return res_int;
}

JsonArray* json_get_array(JsonNode *json_node, const gchar *path)
{
        JsonArray *res_arr = NULL;
        JsonNode *result = json_path_query(path, json_node, NULL);
        if (result) {
                JsonArray *arr = json_node_get_array(result);
                if (json_array_get_length(arr) > 0) {
                        JsonNode *node = json_array_get_element(arr, 0);
                        if (JSON_NODE_HOLDS_ARRAY(node)) {
                                //g_debug("json_get_array: %s\n", json_to_string(result, TRUE));
                                res_arr = json_node_get_array(node);
                        }
                }
        }
        json_node_unref(result);
        return res_arr;
}

gboolean json_contains(JsonNode *root, const gchar *key)
{
        JsonNode *node = json_path_query(key, root, NULL);
        gboolean result = (node != NULL && json_array_get_length(json_node_get_array(node)) > 0);
        json_node_unref(node);
        return result;
}


include/json-helper.h
/**
 * SPDX-License-Identifier: LGPL-2.1-only
 *
 * Copyright (C) 2018-2020 Prevas A/S (www.prevas.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#ifndef __JSON_HELPER_H__
#define __JSON_HELPER_H__

#include <glib-2.0/glib.h>
#include <json-glib/json-glib.h>

gchar* json_get_string(JsonNode *json_node, const gchar *path);
gint64 json_get_int(JsonNode *json_node, const gchar *path);
JsonArray* json_get_array(JsonNode *json_node, const gchar *path);
gboolean json_contains(JsonNode *root, const gchar *key);

#endif // __JSON_HELPER_H__


src/config-file.c
/**
 * SPDX-License-Identifier: LGPL-2.1-only
 *
 * Copyright (C) 2018-2020 Prevas A/S (www.prevas.com)
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @file config-file.c
 * @author Lasse Mikkelsen <lkmi@prevas.dk>
 * @date 19 Sep 2018
 * @brief Configuration file parser
 *
 */

#include "config-file.h"

static const gint DEFAULT_CONNECTTIMEOUT  = 20;     // 20 sec.
static const gint DEFAULT_TIMEOUT         = 60;     // 1 min.
static const gint DEFAULT_RETRY_WAIT      = 5 * 60; // 5 min.
static const gboolean DEFAULT_SSL         = TRUE;
static const gboolean DEFAULT_SSL_VERIFY  = TRUE;
static const gchar * DEFAULT_LOG_LEVEL    = "message";

void config_file_free(struct config *config)
{
        g_free(config->hawkbit_server);
        g_free(config->controller_id);
        g_free(config->tenant_id);
        g_free(config->auth_token);
        g_free(config->gateway_token);
        g_free(config->bundle_download_location);
        g_hash_table_destroy(config->device);
}

static gboolean get_key_string(GKeyFile *key_file, const gchar* group, const gchar* key, gchar** value, const gchar* default_value, GError **error)
{
        gchar *val = NULL;
        val = g_key_file_get_string(key_file, group, key, NULL);
        if (val == NULL) {
                if (default_value != NULL) {
                        *value = g_strdup(default_value);
                        return TRUE;
                }

                g_set_error(error, G_KEY_FILE_ERROR,
                            G_KEY_FILE_ERROR_NOT_FOUND,
                            "Key '%s' not found in group '%s' and no default given", key, group);
                return FALSE;
        }
        *value = val;
        return TRUE;
}

static gboolean get_key_bool(GKeyFile *key_file, const gchar* group, const gchar* key, gboolean* value, const gboolean default_value, GError **error)
{
        g_autofree gchar *val = NULL;
        gboolean val_false, val_true;
        val = g_key_file_get_string(key_file, group, key, NULL);
        if (val == NULL) {
                *value = default_value;
                return TRUE;
        }
        val_false = (g_strcmp0(val, "0") == 0 || g_ascii_strcasecmp(val, "no") == 0 || g_ascii_strcasecmp(val, "false") == 0);
        if (val_false) {
                *value = FALSE;
                return TRUE;
        }
        val_true = (g_strcmp0(val, "1") == 0 || g_ascii_strcasecmp(val, "yes") == 0 || g_ascii_strcasecmp(val, "true") == 0);
        if (val_true) {
                *value = TRUE;
                return TRUE;
        }

        g_set_error(error, G_KEY_FILE_ERROR,
                    G_KEY_FILE_ERROR_INVALID_VALUE,
                    "Value '%s' cannot be interpreted as a boolean.", val);

        return FALSE;
}

static gboolean get_key_int(GKeyFile *key_file, const gchar* group, const gchar* key, gint* value, const gint default_value, GError **error)
{
        GError *ierror = NULL;
        gint val = g_key_file_get_integer(key_file, group, key, &ierror);

        if (val == 0 && g_error_matches(ierror, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_KEY_NOT_FOUND)) {
                g_clear_error(&ierror);
                *value = default_value;
                return TRUE;
        }
        else if (val == 0 && ierror) {
                g_propagate_error(error, ierror);
                return FALSE;
        }
        *value = val;
        return TRUE;
}

static gboolean get_group(GKeyFile *key_file, const gchar *group, GHashTable **hash, GError **error)
{
        guint key;
        gsize num_keys;
        gchar **keys;

        *hash = g_hash_table_new(g_str_hash, g_str_equal);
        keys = g_key_file_get_keys(key_file, group, &num_keys, error);
        if (keys == NULL)
                return FALSE;

        if (num_keys == 0) {
                g_set_error(error, G_KEY_FILE_ERROR,
                            G_KEY_FILE_ERROR_PARSE,
                            "Group '%s' has no keys set", group);
                return FALSE;
        }

        for (key = 0; key < num_keys; key++)
        {
                gchar *value = g_key_file_get_value(key_file,
                                                    group,
                                                    keys[key],
                                                    error);
                if (value == NULL)
                        return FALSE;

                g_hash_table_insert(*hash, keys[key], value);
                //g_debug("\t\tkey %u/%lu: \t%s => %s\n", key, num_keys - 1, keys[key], value);
        }

        return TRUE;
}

static GLogLevelFlags log_level_from_string(const gchar *log_level)
{
        if (g_strcmp0(log_level, "error") == 0) {
                return G_LOG_LEVEL_ERROR;
        } else if (g_strcmp0(log_level, "critical") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL;
        } else if (g_strcmp0(log_level, "warning") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING;
        } else if (g_strcmp0(log_level, "message") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE;
        } else if (g_strcmp0(log_level, "info") == 0) {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE |
                       G_LOG_LEVEL_INFO;
        } else if (g_strcmp0(log_level, "debug") == 0) {
                return G_LOG_LEVEL_MASK;
        } else {
                return G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL |
                       G_LOG_LEVEL_WARNING | G_LOG_LEVEL_MESSAGE;
        }
}

struct config* load_config_file(const gchar* config_file, GError** error)
{
        struct config *config = g_new0(struct config, 1);

        gint val_int;
        g_autofree gchar *val = NULL;
        g_autoptr(GKeyFile) ini_file = g_key_file_new();
        gboolean key_auth_token_exists = FALSE;
        gboolean key_gateway_token_exists = FALSE;

        if (!g_key_file_load_from_file(ini_file, config_file, G_KEY_FILE_NONE, error)) {
                return NULL;
        }

        if (!get_key_string(ini_file, "client", "hawkbit_server", &config->hawkbit_server, NULL, error))
                return NULL;

        key_auth_token_exists = get_key_string(ini_file, "client", "auth_token", &config->auth_token, NULL, NULL);
        key_gateway_token_exists = get_key_string(ini_file, "client", "gateway_token", &config->gateway_token, NULL, NULL);
        if (!key_auth_token_exists && !key_gateway_token_exists) {
                g_set_error(error, 1, 4, "Neither auth_token nor gateway_token is set in the config.");
                return NULL;
        } else if (key_auth_token_exists && key_gateway_token_exists) {
                g_warning("Both auth_token and gateway_token are set in the config.");
        }

        if (!get_key_string(ini_file, "client", "target_name", &config->controller_id, NULL, error))
                return NULL;
        if (!get_key_string(ini_file, "client", "tenant_id", &config->tenant_id, "DEFAULT", error))
                return NULL;
        if (!get_key_string(ini_file, "client", "bundle_download_location", &config->bundle_download_location, NULL, error))
                return NULL;
        if (!get_key_bool(ini_file, "client", "ssl", &config->ssl, DEFAULT_SSL, error))
                return NULL;
        if (!get_key_bool(ini_file, "client", "ssl_verify", &config->ssl_verify, DEFAULT_SSL_VERIFY, error))
                return NULL;
        if (!get_group(ini_file, "device", &config->device, error))
                return NULL;

        if (!get_key_int(ini_file, "client", "connect_timeout", &val_int, DEFAULT_CONNECTTIMEOUT, error))
                return NULL;
        config->connect_timeout = val_int;

        if (!get_key_int(ini_file, "client", "timeout", &val_int, DEFAULT_TIMEOUT, error))
                return NULL;
        config->timeout = val_int;

        if (!get_key_int(ini_file, "client", "retry_wait", &val_int, DEFAULT_RETRY_WAIT, error))
                return NULL;
        config->retry_wait = val_int;

        if (!get_key_string(ini_file, "client", "log_level", &val, DEFAULT_LOG_LEVEL, error))
                return NULL;
        config->log_level = log_level_from_string(val);

        if (config->timeout > 0 && config->connect_timeout > 0 && config->timeout < config->connect_timeout) {
                g_set_error(error,
                            G_KEY_FILE_ERROR,                   // error domain
                            G_KEY_FILE_ERROR_INVALID_VALUE,     // error code
                            "timeout should be greater than connect_timeout. Timeout: %ld, Connect timeout: %ld",
                            config->timeout,
                            config->connect_timeout
                            );
                return NULL;
        }

        return config;
}
